src/context/json.test.ts

```diff
@@ -14,7 +14,7 @@ test('sets given Array as the response JSOn body', () => {
 
 test('sets given string as the response JSON body', () => {
   const result = response(json('some string'))
-  expect(result).toHaveProperty('body', `"some string"`)
+  expect(result).toHaveProperty('body', 'some string')
```

Comment:
Do we drop the extra quotes assuming the body is going to be stringified later on by the default response transformer?

Reply:
Hi @kettanaito, the `json` function does not stringify the content, the responsible to do that is the response transformer. In this case the default response transformer.

The difference is that the default transformer checks the content is already a string so it does not stringify it again (cause of the extra quotes).

Reply:
Can we ensure that parsing of such JSON response body works as intended?

For example, `JSON.parse('some string')` would throw an exception as the given string indeed lacks those extra quotes. Since here we are asserting on the `response()` invocation result, that's what the consumer may get. 

There is no additional response transformation since the response composition moment: body gets transferred to the worker script and a `Response` instance is constructed with that body:

```js
new res = new Response('some string')
await res.json()
```

The example above will throw a JSON parsing exception as the body string misses extra quotes around it to be a valid JSON.

Reply:
Hi @kettanaito I've got you point and it makes sense but I am not sure how to make the change though. 

To be honest I am not that familiar with `msw` internals so I was reading the code and the past PRs and trying to figure that out. I saw that PR #403 changed the `ctx.json` which, as expected, was the responsible to stringify the body content. Now, it's responsible to set the header `content-type: application/json` and enable merges. 

To me the merge functionality seems like something we could move to a utils function so when calling `json`, `data` or `errors` functions we could have something like:
```js
ctx.json(merge(data1, data2, data3))
```
It seems the project already have a function like this (`mergeRight`) so it was just a matter of export the function and update the docs.

Also, correct me if I am wrong, but it seems the response transformers were added just to enable different json parsers in the body (#440). If the merge was a util/helper function we could solve the problem just defining a custom resolve transformer like:
```js
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import * as JSONbig from 'json-bigint'

const jsonBigInt = body => {
  return (res) => {
    res.headers.set('Content-Type', 'application/json')
    res.body = JSONbig.stringify(body)
    return res
  }
}

const server = setupServer(
  rest.get('http://test.mswjs.io/me', (req, res, ctx) => {
    const me = {
      username: 'Dude',
      balance: BigInt(1597928668063727616),
    }
    return res(ctx.jsonBigInt(me))
  }),
)
```

This pattern can be use to extend the available context functions for different use cases. This is what I do in my projects.

I know is not that simple to change the API of a open source library so as an alternative solution we could create a flag in the default response:

```ts
export interface MockedResponse<BodyType = any> {
  body: BodyType
  status: number
  statusText: string
  headers: Headers
  once: boolean
  isJson: boolean // new flag to point to json responses
  delay?: number
}
```

The default would be **false** and we would set it to **true** in the `json` function. Then we could check for the flag in the default response transformer like:
```ts
const stringifyJsonBody: ResponseTransformer = (res) => {
  if (res.body && res.isJson) {
    res.body = JSON.stringify(res.body)
  }
  return res
}
```

Notice we don't need to check for the header anymore. We only guarantee the body to be stringified if the `ctx.json` function is used. In case the users set the content-type header with any json variation manually and use `ctx.text`  or `ctx.body` to set the body, the user would need to stringify the body themselves.


I personally try to avoid flags as much as I can because they add complexity.

Let me know what you think.

Reply:
It's a little different than that. We are exposing an explicit set of context utilities depending on the request handler (rest/graphql). That way your mocks resemble a spec-compliant response composition respective to the handler used. For instance, in `graphql.*` there's no `ctx.json` utility, as you cannot set an arbitrary JSON as a GraphQL endpoint response, it must follow the spec. 

That was the reason #403 got introduced. Under the hood it meant an internal extension of the `ctx.json` utility that powers `ctx.data` and `ctx.errors` under the hood, to merge all the given objects of the response body. During the implementation we realized that in order to operate with those objects in a more comfortable fashion they should be stored as objects, not strings. Thus stringification of the JSON response body was moved to be the last step of the response composition function.

When it comes to this test, I believe this becomes a breaking change, as the response body is not a valid JSON. 

> If the merge was a util/helper function we could solve the problem just defining a custom resolve transformer like:

This is a good point. I like the idea of being able to incorporate certain fixed logic to the response composition function itself, but such custom response transformer also satisfies the use case. We should seriously consider this.

> I personally try to avoid flags as much as I can because they add complexity.

I couldn't agree more. I don't think that extending the mocked response object is the way to go. 

Do you think something like a custom default response transformer (like the current JSON BigInt) has a place to be? Or can any use case be substituted by a custom transformer?

Reply:
Regarding the `isJson` property, I find it confusing, as this information should be extracted from the response's headers, treating them as the source of truth. To brainstorm, here are the factors we have:

- `ctx.body` should set a given body as-is, not bothering with its stringification.
- `ctx.json` must always result into a valid JSON body (stringified).
- `ctx.json` may potentially support cross-transformer merges. This item may be obsolete if we fully replace default response transformers with custom `ctx.*` functions developers may create.
- `typeof res.body === 'string` is insufficient to understand whether the response body should be stringified or not. Strings are also a valid JSON.

Reply:
@msutkowski, hey, Matt 👋 What do you think about the points above?